[FrameworkBundle][Routing] added Configurators to handle template and redirect controllers

HeahDude opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | let's see
| Fixed tickets | partially #24640, #25145
| License       | MIT
| Doc PR        | symfony/symfony-docs#11120

While working on https://github.com/symfony/symfony-docs/pull/11085, I felt bad about the long notations required for simple [redirects](https://symfony.com/doc/current/routing/redirect_in_config.html) and [templates rendering](https://symfony.com/doc/current/templating/render_without_controller.html) template actions, but I love and use those features since always. Then I gave it a try yesterday night and now I realised I missed #24640 and that #25145 has been closed x).

So here we go, here's my WIP. WDYT of this implementation? ping @javiereguiluz?

I'm going to open the PR in the docs so we can discuss the DX changes there too, and keep focus on the code here.

Cheers!

EDIT
----
This PR now only update PHP-DSL configurators.

______________

TODO:

- [x] gather reviews
- ~[x] fix xml schema~
- [x] add some tests
- ~[ ] handle xsd auto discovery~
- [x] rebase on top of #30507
- [x] ~add shortcuts for #30514~

HeahDude commented:
I'd really appreciate some help if anyone had already overridden some xsd schema thanks!

HeahDude commented:
I've updated symfony/symfony-docs#11120 with a nice before/after view.

Also I need to rebase on top of #30507.

Xml config is working properly, but the IDE cannot resolve validation/autocompletion yet, this works thanks to the loader using its constant local path.
I guess we need to create a new path on `https://symfony.com/routing` that we need to declare explicitly in `config/routes.xml` or any xml file where this feature should be used, are you the one handling this @fabpot? Can you help me on this?

Apart this last points, the feature is ready from side, only reviews needed now :), thanks!

javiereguiluz commented:
Thanks for working on this. To make this feature completely self-explanatory, I think we should rename `redirect-to` as `redirect-to-route` (we already have `redirect-to-url`). What do you think?

HeahDude committed:
[FrameworkBundle][Routing] added Configurators to handle template and redirect controllers

HeahDude commented:
Rebased and squashed here, while I addressed @javiereguiluz's comment.

Also, I was not happy with the xml definition because all attributes are there and it's hard to guess which one to use  in which context, so I propose to use dedicated types instead: https://github.com/symfony/symfony/compare/a46c76a8c3a05adb7f984ea4d9cffbd3a446f3f5..7055ba0a5b09cec84d88f5c0447e79b65aa48a85#diff-96d406b9b38189be981bbea1bad334bd.

Tests are passing locally and on travis, but not on windows.. I wonder how this line https://github.com/symfony/symfony/pull/30501/files#diff-44f544990841fe196e4ea5c76dcc18dfR8 can be a problem, if anyone has a clue on that topic that would be awesome.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Routing/Loader/PhpFileLoader.php:
```diff
@@ -0,0 +1,28 @@
+{
+    protected function getConfigurator(RouteCollection $collection, string $path, string $file): BaseRoutingConfigurator
+    {
+        return new RoutingConfigurator($collection, $this, $path, $file);
```
this class should be put in the `@return annotation` so that type checkers can know about it

nicolas-grekas commented:
Did you consider this instead for the PHP-DSL?
```php
    $routes->add('template_route', '/static')
        ->template(''static.html.twig')
        ->maxAge(300)
        ->sharedMaxAge(100)
        ->private(true)
        ->methods(['GET'])
        ->options(['utf8' => true])
        ->condition('abc')
    ;
    $routes->add('redirect_route', '/redirect')
        ->redirectToRoute('target_route')
        ->permanent(true)
        ->ignoreAttributes(['attr', 'ibutes'])
        ->keepRequestMethod(true)
        ->keepQueryParams(true)
        ->schemes(['http'])
        ->host('legacy')
        ->options(['utf8' => true])
    ;
    $routes->add('url_redirect_route', '/redirect-url')
        ->redirectToUrl('/url-target')
        ->permanent(true)
        ->scheme('http')
        ->httpPort(1)
        ->httpsPort(2)
        ->keepRequestMethod(true)
        ->host('legacy')
        ->options(['utf8' => true])
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Routing/Loader/YamlFileLoader.php:
```diff
@@ -0,0 +1,82 @@
+    protected function validate($config, $name, $path)
+    {
+        // remove invalid keys for values normalized below
+        foreach (array_unique(array_merge(...array_values(self::$availableKeys))) as $key) {
```
can we validate the attributes per type of route?
can we validate that the routes using one and only one of the 3 possibles types?

HeahDude replied:
When I look at how the parent is implemented and extended, the only valuable thing I can think of checking here is about defining both "controller" and one of the three types we introduce.

HeahDude replied:
We shouldn't need such check since the xsd is already doing the job.

HeahDude replied:
I see now that we could do both what you said and enforcing not using the controller key as I proposed. I'll add the proper tests to cover it in Yaml; the XSD is already doing the job, and the PHP fluent interface does not allow invalid calls by design.

HeahDude replied:
Done in #35653. Also added a new trait to prevent calling `controller` with configurators.

nicolas-grekas commented:
Now that https://github.com/symfony/symfony/pull/34873 is merged, can you please rebase and take it into account to inject your new ContainerConfigurator?

HeahDude commented:
Done, rebased and squashed.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Routing/Loader/Configurator/RouteConfigurator.php:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * @param string $template The template name
+     */
+    final public function template(string $template): TemplateRouteConfigurator
```
will need to account for https://github.com/symfony/symfony/pull/35257 :)

nicolas-grekas replied:
#35257 merged, can be rebased now.

HeahDude replied:
Done.

HeahDude commented:
I've updated symfony/symfony-docs#11120 accordingly.

nicolas-grekas commented on src/Symfony/Component/Routing/Loader/PhpFileLoader.php:
```diff
@@ -48,7 +50,7 @@ public function load($file, string $type = null)
         if (\is_object($result) && \is_callable($result)) {
             $collection = new RouteCollection();
-            $result(new RoutingConfigurator($collection, $this, $path, $file));
+            $result($this->getConfigurator($collection, $path, $file));
```
I'd suggest making the new method responsible for calling $result:
let's do `$this->callConfigurator($result, $collection, $path, $file)` instead?

HeahDude replied:
I've done `$collection = $this->callConfigurator($result, $path, $file);` instead.

nicolas-grekas reviewed (commented):
This looks really good to me, just a minor note.
Please change the type hint in `MicroKernelTrait` also.
This will lead to a recipe update.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/routing-1.0.xsd:
```diff
@@ -0,0 +1,86 @@
+            targetNamespace="http://symfony.com/schema/routing"
+            elementFormDefault="qualified">
+
+  <xsd:include schemaLocation="http://symfony.com/schema/routing/routing-1.0.xsd" />
```
where is the new extended routing scheme going to be hosted? a new location means all the documentation would need updates.

HeahDude replied:
Split in #35653.

Tobion reviewed (commented):
I like the configurator solution and the xml and yaml solution seems fine to me too. But I wonder if we really need to add the custom options for xml and yaml. It means IDEs like phpstorm explicitly need to add support for them. Considering we seem to default to the php configurator for apps in the future (https://github.com/symfony/recipes/pull/721), maybe we only need to implement this for the php configurator? If people want to use yaml or xml in bundles, they can still do so using the existing notation which does not need updates to IDEs and documentation.

HeahDude commented:
Thank you @Tobion for the review. I agree we could only support PHP-fluent, easier to implement and document, and maybe I can split my PR in two or discard others formats.

However, about:
>But I wonder if we really need to add the custom options for xml and yaml. It means IDEs like phpstorm explicitly need to add support for them.

That's only true for YAML which would need a plugin update indeed. But the XSD is here to ensure that auto-completion and validation are mostly "native" in many tools too.

>where is the new extended routing scheme going to be hosted? a new location means all the documentation would need updates.

I guess we should publish a new path like `http://symfony.com/schema/routing/framework-routing-1.0.xsd`. We could either document it as part of the features only (a few example in the docs) by adding a note mentioning that they require this specific schema. Or do a find/replace in all the docs, it may worth it.

Also, I think we should consider that the related issue is about a YAML feature request, I guess because for now YAML is the first tab we show in the docs, so ppl may miss this feature, PHP being still the third and last tab for now. Maybe we should change that visibility first before making this feature exclusive to PHP configuration?

Tobion commented:
> maybe I can split my PR in two or discard others formats

I would like it. Then we can decide later if yaml and xml is needed and how to solve the documentation and xsd path issues.

HeahDude commented:
I've updated this PR to only update the PHP loader. ping @Tobion @nicolas-grekas

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Routing/Loader/Configurator/Traits/RouteTrait.php:
```diff
@@ -0,0 +1,22 @@
+
+use Symfony\Component\Routing\Loader\Configurator\RouteConfigurator;
+
+trait RouteTrait
```
what about naming this NoControllerTrait?

HeahDude replied:
Good call!

HeahDude replied:
Removed after our discussion on Slack.

nicolas-grekas commented:
Thank you @HeahDude.

Tobion commented on src/Symfony/Component/Routing/Loader/Configurator/Traits/LocalizedRouteTrait.php:
```diff
@@ -0,0 +1,74 @@
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+trait LocalizedRouteTrait
```
shouldn't all these traits be marked as internal?

HeahDude replied:
There are used in the framework bundle and should in any extension. I guess not

Tobion replied:
yes but the implementation as a trait is an implementation detail and the trait should not be reused by users

HeahDude replied:
I think they should if they need to add custom loaders/configurators as done here

Tobion replied:
If users add custom configurators, they would use different method with other return types (like this PR shows) which would not use those traits but only the classes.

HeahDude replied:
Yes I agree that it would be better to extend configurators that use the traits directly. But if one want to create a `JsonLoader` or any other type it would be nice to use the traits as done here for YAML and XML but the I created this trait to also "reuse" its method in the new configurator. Maybe we could have an `AbstractRoutingLoader` using the traits and extended by the YAML and XML one, so users can extend it too instead or using the traits directly?

HeahDude replied:
@Tobion see #35731.
